pkg/registry-disk/registry-disk.go

```diff
@@ -0,0 +1,197 @@
+/*
+ * This file is part of the kubevirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package registrydisk
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/jeevatkm/go-model"
+
+	kubev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+)
+
+const registryDiskV1Alpha = "ContainerRegistryDisk:v1alpha"
+const defaultIqn = "iqn.2017-01.io.kubevirt:wrapper/1"
+
+func DisksAreReady(pod *kubev1.Pod) bool {
+	// Wait for readiness probes on image wrapper containers
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		if strings.Contains(containerStatus.Name, "disk") == false {
+			// only check readiness of disk containers
+			continue
+		}
+		if containerStatus.Ready == false {
+			return false
+		}
+	}
+	return true
+}
+
+// The virt-handler converts registry disks to their corresponding iscsi network
+// disks when the VM spec is being defined as a domain with libvirt.
+// The ports and host of the iscsi disks are already provided here by the controller.
+func MapRegistryDisks(vm *v1.VM) (*v1.VM, error) {
+	vmCopy := &v1.VM{}
+	model.Copy(vmCopy, vm)
+
+	for idx, disk := range vmCopy.Spec.Domain.Devices.Disks {
+		if disk.Type == registryDiskV1Alpha {
+			newDisk := v1.Disk{}
+
+			newDisk.Type = "network"
+			newDisk.Device = "disk"
+			newDisk.Target = disk.Target
+
+			newDisk.Driver = &v1.DiskDriver{
+				Type:  "raw",
+				Name:  "qemu",
+				Cache: "none",
+			}
+
+			newDisk.Source.Name = defaultIqn
+			newDisk.Source.Protocol = "iscsi"
+			newDisk.Source.Host = disk.Source.Host
+
+			vmCopy.Spec.Domain.Devices.Disks[idx] = newDisk
+		}
+	}
+
+	return vmCopy, nil
+}
+
+// TODO Introduce logic that dynamically generates iscsi CHAP
+// Authentication credentials for a VM spec backed by registry disks.
+//
+//func ApplyAuth(vm *v1.VM) {
+//     INSERT DYNAMIC AUTH LOGIC HERE
+//}
+
+// The controller applies ports to registry disks when a VM spec is introduced into the cluster.
+func ApplyPorts(vm *v1.VM) {
+	wrapperStartingPort := 3261
+	for idx, disk := range vm.Spec.Domain.Devices.Disks {
+		if disk.Type == registryDiskV1Alpha {
+			port := fmt.Sprintf("%d", wrapperStartingPort)
+			name := "127.0.0.1"
```

Comment:
127.0.0.1 is the default host until it is overwritten by the ip associated with the scheduled pod. In my update I'm making this more clear.